staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +190,206 @@ func RecoverFromPanic(err *error) {
+	// TODO remove this once we have a useful default value for similar
+	if d.Similar >= 1 {
+		return errVal{}, true // return true because this is a new error
+	}
+
+	var (
+		// fuzzySimilarity is initialized to d.similar since that is the
+		// lowest similarity value we consider as the strings being "equal"
+		fuzzySimilarity = d.Similar
+		fuzzyFound      bool
+		fuzzyMessage    string
+		fuzzyVal        errVal
+	)
+
+	// we have to iterate over the whole map to do fuzzy matching on errKey.message
+	// this is ok because d.count should always be relatively small
+	// note that we cannot exit this loop early since we want to pick the most similar message
+	for k, v := range d.count {
+		// the stack and error type must match
+		if key.stack == k.stack && key.errType == k.errType {
+			// now we perform fuzzy matching on the message
+			if s := levenshtein.Similarity(key.message, k.message, nil); s > fuzzySimilarity {
+				fuzzySimilarity = s
+				fuzzyFound = true
+				fuzzyMessage = k.message
+				fuzzyVal = v
+			}
+		}
+	}
+
+	if fuzzyFound {
+		// update the input key's message to match our fuzzy search
+		key.message = fuzzyMessage
+		return fuzzyVal, false // return false because we do not consider this a new error
+	}
+
+	return errVal{}, true // return true because this is a new error
+}
+
+// logError uses glog to log at the call site of HandleError
+// it must be called from dedupingErrorHandler.handleErr
+func (d *dedupingErrorHandler) logError(err error, count uint64) {
+	// get the full name of the file since glog likes to strip it
+	// TODO if we are ok with multi line messages, we could simply print the whole stack
+	_, file, line, ok := runtime.Caller(d.errorDepth)
+	if !ok {
+		file = "unknown_file"
+		line = -1
+	}
+	glog.ErrorDepth(d.errorDepth, fmt.Sprintf("%s:%d err=%v count=%d value=%#v", file, line, err, count, err))
+}
+
+var (
+	hexNumberRE  = regexp.MustCompile(`0x[0-9a-f]+`)
+	emptyAddress = []byte("0x?")
+)
+
+// getStack returns the important part of the stack trace
+// it must be called from dedupingErrorHandler.handleErr
+// it must not mutate d (or even read d.cache/d.count) since it is called when the lock is not held
+func (d *dedupingErrorHandler) getStack() string {
+	// remove all hex addresses from the stack dump because closures can have volatile values
+	stack := string(hexNumberRE.ReplaceAll(debug.Stack(), emptyAddress))
+	// strip the redundant stuff at the top of the stack
+	// add 1 to error depth for debug.Stack (since it calls runtime.Stack), times the sum by 2 since each frame has 2 lines
+	// add 1 for go routine number header
+	strip := (d.errorDepth+1)*2 + 1
+	stackLines := strings.Split(stack, "\n")
+	// do not panic trying to strip a stack trace that does not meet our expectations
+	if strip >= len(stackLines) {
+		return stack
+	}
+	return strings.Join(stackLines[strip:], "\n")
+}
+
+func isPowerOfTwo(n uint64) bool {
+	return (n & (n - 1)) == 0
```

Comment:
I think this claims that 0 is a power of 2?

Reply:
It is only used for values > 0, and it would be fine even if it did get used for 0